sdv/metadata/multi_table.py

```diff
@@ -13,6 +15,165 @@ def __init__(self):
+        """Validate that there is no circular relatioship in the metadata."""
+        parents = set() if parents is None else parents
+        if children is None:
+            children = child_map[parent]
+
+        if parent in children:
+            errors.append(parent)
+
+        for child in children:
+            if child in parents:
+                break
+
+            parents.add(child)
+            self._validate_circular_relationships(
+                parent,
+                children=child_map.get(child, set()),
+                child_map=child_map,
+                parents=parents,
+                errors=errors
+            )
+
+    def add_relationship(self, parent_table_name, child_table_name,
+                         parent_primary_key, child_foreign_key):
+        """Add a relationship between two tables.
+
+        Args:
+            parent_table_name (str):
+                A string representing the name of the parent table.
+            child_table_name (str):
+                A string representing the name of the child table.
+            parent_primary_key (str or tuple):
+                A string or tuple of strings representing the primary key of the parent.
+            child_foreign_key (str or tuple):
+                A string or tuple of strings representing the foreign key of the child.
+
+        Raises:
+            - ``ValueError`` if a table is missing.
+            - ``ValueError`` if the ``parent_primary_key`` or ``child_foreign_key`` are missing.
+            - ``ValueError`` if the ``parent_primary_key`` and ``child_foreign_key`` have different
+              size.
+            - ``ValueError`` if the ``parent_primary_key`` and ``child_foreign_key`` are different
+              ``sdtype``.
+            - ``ValueError`` if the relationship causes a circular dependency.
+        """
+        parent_table = self._tables.get(parent_table_name)
+        self._validate_missing_relationship_tables(
+            parent_table_name, child_table_name, self._tables.keys())
+
+        self._validate_missing_relationship_keys(
+            parent_table,
+            parent_table_name,
+            parent_primary_key,
+            child_table_name,
+            child_foreign_key
+        )
+        self._validate_relationship_key_length(
+            parent_table_name, parent_primary_key, child_table_name, child_foreign_key)
+
+        self._validate_relationship_sdtypes(
+            parent_table,
+            parent_table_name,
+            parent_primary_key,
+            child_table_name,
+            child_foreign_key
+        )
+
+        child_map = defaultdict(set)
+        for relation in self._relationships:
+            parent_name = relation['parent_table_name']
+            child_name = relation['child_table_name']
+            child_map[parent_name].add(child_name)
+
+        child_map[parent_table_name].add(child_table_name)
+        errors = []
```

Comment:
Can you move this logic (160-168) to its own method? I think it makes this method easier to read, makes this logic easier to test and harder for us to mess this up in the future (since you are mutating errors inplace I'd prefer to be extra careful).